stan/math/prim/eigen_plugins.h

```diff
@@ -51,29 +69,28 @@ struct val_Op{
   EIGEN_EMPTY_STRUCT_CTOR(val_Op);
 
   //Returns value from a vari*
-  template<typename T = Scalar>
+  template <typename T = Scalar,
+           std::enable_if_t<is_vari<T>::value>* = nullptr>
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
-    std::enable_if_t<std::is_pointer<T>::value, const double&>
-      operator()(T &v) const { return v->val_; }
+    const double& operator()(T& v) const { return v->val_; }
 
   //Returns value from a var
-  template<typename T = Scalar>
+  template <typename T = Scalar,
+           std::enable_if_t<is_var<T>::value>* = nullptr>
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
-    std::enable_if_t<(!std::is_pointer<T>::value && !is_fvar<T>::value
-                      && !std::is_arithmetic<T>::value), const double&>
-      operator()(T &v) const { return v.vi_->val_; }
+    const double& operator()(T& v) const { return v.vi_->val_; }
 
   //Returns value from an fvar
-  template<typename T = Scalar>
+  template <typename T = Scalar,
+           std::enable_if_t<is_fvar<T>::value>* = nullptr>
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
-    std::enable_if_t<is_fvar<T>::value, forward_return_t<T>>
-      operator()(T &v) const { return v.val_; }
+    forward_return_t<T> operator()(T& v) const { return v.val_; }
 
   //Returns double unchanged from input (by value)
-  template<typename T = Scalar>
+  template <typename T = Scalar,
+           std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
-    std::enable_if_t<std::is_arithmetic<T>::value, double_return_t<T>>
-      operator()(T v) const { return v; }
+    double_return_t<T> operator()(T v) const { return v; }
```

Comment:
Why is this one just plain `T`?

Reply:
I'm not actually sure why this is here, I'll remove and see if the tests complain

Reply:
A double is 8 bytes and so is a ref (since refs are really just pointers from the compilers viewpoint). But I think I'd make these signatures `T&&` just so that `double_return_t<T>` returns back the right cv qualified type

Reply:
Though eigen doesn't complain right now so maybe it's never supposed to return a `const double`? Or could this be where the bug is digging up that requires `val_op()`?